synapse/metrics/__init__.py

```diff
@@ -575,11 +582,16 @@ def f(*args, **kwargs):
             return ret
 
         # Check if we need to do a manual GC (since its been disabled), and do
-        # one if necessary.
+        # one if necessary. Note we go in reverse order as e.g. a gen 1 GC may
+        # promote an object into gen 2, and we don't want to handle the same
+        # object multiple times.
         threshold = gc.get_threshold()
         counts = gc.get_count()
         for i in (2, 1, 0):
-            if threshold[i] < counts[i]:
+            # We check if we need to do one based on a straightforward
+            # comparison between the threshold and count. We also do an extra
+            # check to make sure that we don't a GC too often.
+            if threshold[i] < counts[i] and MIN_TIME_BETWEEN_GCS[i] < end - _last_gc[i]:
```

Comment:
This can become out-of-bounds if someone specifies `[1, 10]` or similar accidentally.

```suggestion
            if threshold[i] < counts[i] and len(MIN_TIME_BETWEEN_GCS) => i and MIN_TIME_BETWEEN_GCS[i] < end - _last_gc[i]:
```

I'm also wary of mutating a "constant" like this, because it can give mixed signals in the code.

Reply:
Yeah, mutating is a bit icky, but I can't think of a better way of injecting the values at this point.

> This can become out-of-bounds if someone specifies [1, 10] or similar accidentally.

We check for the right format in the config, so I think its fine that this explodes if a developer messes up and sets it incorrectly.